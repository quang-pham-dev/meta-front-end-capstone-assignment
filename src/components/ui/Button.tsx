import React from "react";

type ButtonVariant = "default" | "primary" | "secondary" | "outline";
type ButtonSize = "small" | "medium" | "large";

interface ButtonProps extends React.ButtonHTMLAttributes<HTMLButtonElement> {
  variant?: ButtonVariant;
  size?: ButtonSize;
}

const Button = React.forwardRef<HTMLButtonElement, ButtonProps>(
  (
    { className, variant = "default", size = "medium", children, ...props },
    ref
  ) => {
    const classes = [
      "button",
      `button--${variant}`,
      `button--${size}`,
      className
    ]
      .filter(Boolean)
      .join(" ");

    return (
      <button className={classes} ref={ref} {...props}>
        {children}
      </button>
    );
  }
);

Button.displayName = "Button";

export { Button };
