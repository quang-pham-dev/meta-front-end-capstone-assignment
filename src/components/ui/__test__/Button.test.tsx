import { render, screen } from "@testing-library/react";
import { describe, it, expect } from "vitest";
import "@testing-library/jest-dom";

import { Button } from "../Button";
import "@testing-library/jest-dom";

describe("Button", () => {
  it("renders with default props", () => {
    render(<Button>Click me</Button>);
    const button = screen.getByRole("button", { name: /click me/i });
    expect(button).toBeInTheDocument();
    expect(button).toHaveClass("button button--default button--medium");
  });

  it("renders with custom variant and size", () => {
    render(
      <Button variant="primary" size="large">
        Custom Button
      </Button>
    );
    const button = screen.getByRole("button", { name: /custom button/i });
    expect(button).toHaveClass("button button--primary button--large");
  });

  it("passes additional props to the button element", () => {
    render(<Button disabled>Disabled Button</Button>);
    const button = screen.getByRole("button", { name: /disabled button/i });
    expect(button).toBeDisabled();
  });

  it("matches snapshot", () => {
    const { asFragment } = render(<Button>Snapshot Button</Button>);
    expect(asFragment()).toMatchSnapshot();
  });
});
