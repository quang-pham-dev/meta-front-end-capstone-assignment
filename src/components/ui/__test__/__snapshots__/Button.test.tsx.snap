// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Button > matches snapshot 1`] = `
<DocumentFragment>
  <button
    class="button button--default button--medium"
  >
    Snapshot Button
  </button>
</DocumentFragment>
`;
